packages/firestore/test/unit/local/web_storage_metadata_channel.test.ts

```diff
@@ -0,0 +1,271 @@
+    });
+
+    it('with empty targets', () => {
+      assertInstanceState([]);
+    });
+
+    it('with multiple targets', () => {
+      notificationChannel.addLocallyActiveQueryTarget(0);
+      assertInstanceState([0]);
+
+      notificationChannel.addLocallyActiveQueryTarget(1);
+      notificationChannel.addLocallyActiveQueryTarget(2);
+      assertInstanceState([0, 1, 2]);
+
+      notificationChannel.removeLocallyActiveQueryTarget(1);
+      assertInstanceState([0, 2]);
+    });
+  });
+
+  describe('combines instance data', () => {
+    let previousAddEventListener;
+    let storageCallback: (StorageEvent) => void;
+
+    beforeEach(() => {
+      previousAddEventListener = window.addEventListener;
+      window.addEventListener = (type, callback) => {
+        expect(type).to.be.equal('storage');
+        storageCallback = callback;
+      };
+
+      return persistenceHelpers
+        .testLocalStorageNotificationChannel(ownerId, [1, 2], [3, 4])
+        .then(nc => {
+          notificationChannel = nc;
+          expect(storageCallback).to.not.be.undefined;
+        });
+    });
+
+    afterEach(() => {
+      notificationChannel.shutdown();
+      window.addEventListener = previousAddEventListener;
+    });
+
+    const verifyState = (minBatchId: BatchId, expectedTargets: TargetId[]) => {
+      const actualTargets = notificationChannel.getGloballyActiveQueryTargets();
+      expect(actualTargets.size).to.equal(expectedTargets.length);
+
+      let missingTargets = actualTargets;
+      for (const targetId of expectedTargets) {
+        missingTargets = missingTargets.delete(targetId);
+      }
+      expect(missingTargets.size).to.equal(0);
+
+      expect(
+        notificationChannel.getMinimumGloballyPendingMutation()
+      ).to.be.equal(minBatchId);
+    };
+
+    it('with existing instance', () => {
+      // The prior instance has one pending mutation and two active query targets
+      verifyState(1, [3, 4]);
+
+      notificationChannel.addLocalPendingMutation(3);
+      notificationChannel.addLocallyActiveQueryTarget(4);
+      verifyState(1, [3, 4]);
+
+      notificationChannel.addLocalPendingMutation(0);
+      notificationChannel.addLocallyActiveQueryTarget(5);
+      verifyState(0, [3, 4, 5]);
+
+      notificationChannel.removeLocalPendingMutation(0);
+      notificationChannel.removeLocallyActiveQueryTarget(5);
+      verifyState(1, [3, 4]);
+    });
+
+    it('from new instances', () => {
+      const secondaryInstanceKey = `fs_instances_${
+        persistenceHelpers.TEST_PERSISTENCE_PREFIX
+      }_${AutoId.newId()}`;
+
+      const oldState: InstanceStateSchema = {
+        lastUpdateTime: Date.now(),
+        activeTargetIds: [5],
+        minMutationBatchId: 0,
+        maxMutationBatchId: 0
+      };
+
+      const updatedState: InstanceStateSchema = {
+        lastUpdateTime: Date.now(),
+        activeTargetIds: [5, 6],
+        minMutationBatchId: 0,
+        maxMutationBatchId: 0
+      };
```

Comment:
I think I'd be much happier to see us actually creating multiple channels and testing that modifications via one are surfaced correctly in others.  As-is, we're skipping a big chunk of the implementation (LocalStorage) and making our tests dependent on implementation details (the LocalStorage schema) unnecessarily.

That would also mean that if we implemented persistence on Node one day and used a communication channel other than LocalStorage, this testing would cover it, so it's more valuable.

Reply:
Some of this testing exists hidden in `testWebStorageSharedClientState`. I updated the test here to rely on the schema embedded in RemoteClientState, but as pointed out before, I can't actually test the notifications.